File: src/main/clojure/comzotohcljc/wflow/pipeline.clj

(ns ^{ doc: ""
       :author "kenl" }
  comzotohcljc.wflow.sysobjs )


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defprotocol JobAPI
  (container [_] ))

(deftype Job

(defprotocol PipelineAPI
  (container [_] )
  (isActive [_] )
  (onError [_ error curPoint] )
  (start [_] )
  (stop [_] ))

(deftype Pipeline [job pid ^:unsynchronized-mutable active delegate]

  Object

  (finalize [this]
    (debug (str "=====================>" (toString this) " finz'ed")) )

  (toString [_]
    (str (-> (.getClass)(.getSimpleName)) "(" pid ")" ))

  PipelineAPI

  (container [_] (.container job))

  (isActive [_] active)

  (onError [this err cur]
    (let [ h (:on-error delegate) ]
      (error err)
      (if (nil? h)
        (ac-reify-nihil this)
        (try
          (apply h err cur)
          (catch Throwable e#
            (do (error e#) (ac-reify-nihil this))
            )))))

  (start [this]
    (let [ f1 (ac-reify-nihil this)
           h (:on-start delegate) ]
      (debug "Pipeline " pid " starting...")
      (try
        (let [ a (apply h this)
               f2 (cond
                      (or (nil? a) (satisfies? Nihil a))
                      (ac-reify-nihil this)
                      :else
                      (ac-reify a f1)) ]
        (-> this (.container) (.core) (.run f2))
        (set! active true)
        (catch Throwable e#
          (onError this e f2))))) )

  (stop [_]
    (try
      (debug "Pipeline " pid " stopping...")
      (let [ h (:on-stop delegate) ]
        (when-not (nil? h)
          (apply h)))
      (catch Throwable e# (error e#))))

  )


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


(defn make-pipelie [job delegate]
  (let [ pid (SN/next-long)
         pipe (Pipeline. job pid false delegate) ]
    (debug "Pipeline " pid " created.")
    (SU/test-nonil "the-job" job)))


(defn- no-flow [pipe job]
  (let [ f (fn [_] (ac-reify-nihil pipe)) ]
  (make-pipeline job { :on-start f })) )














(def ^:private sysobjs-eof nil)

File: src/main/scala/com/zotoh/wflow/Pipeline.scala
/*??
*
* Copyright (c) 2013 Cherimoia, LLC. All rights reserved.
*
* This library is distributed in the hope that it will be useful
* but without any warranty; without even the implied warranty of
* merchantability or fitness for a particular purpose.
*
* The use and distribution terms for this software are covered by the
* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
* which can be found in the file epl-v10.html at the root of this distribution.
*
* By using this software in any fashion, you are agreeing to be bound by
* the terms of this license.
* You must not remove this notice, or any other, from this software.
*
 ??*/


package com.zotoh.wflow

import com.zotoh.frwk.util.CoreUtils._
import org.slf4j._
import com.zotoh.frwk.util.Schedulable
import java.util.concurrent.atomic.AtomicLong
import com.zotoh.wflow.core.Scope
import com.zotoh.frwk.core.Startable
import com.zotoh.frwk.server.ServerLike


trait PipelineDelegate {

  def getStartActivity(pipe:Pipeline) : Activity
  def onStop(pipe:Pipeline) : Unit
  def onError(err:Throwable, curPt:FlowPoint) : Activity

}



object Pipeline {
  private val _sn= new AtomicLong(0L)
  private def nextId() = _sn.incrementAndGet()
}

/**
 * @author kenl
 *
 */
class Pipeline (private val _theScope:Scope, private val _delegateClass:String) extends Startable {

  private val _log:Logger= LoggerFactory.getLogger(classOf[Pipeline] )
  def tlog() = _log

  import Pipeline._

  private val _delegate = Thread.currentThread().getContextClassLoader().loadClass(_delegateClass).getDeclaredConstructor().newInstance() match {
    case x:PipelineDelegate => x
    case _ =>
    throw new ClassCastException("Class " + _delegateClass + " must implement PipelineDelegate.")
  }

  private val _pid= nextId()
  private var _active=false

  tlog().debug("{}: {} => pid : {}" , "Pipeline", getClass().getName() , asJObj(_pid))
  require(_theScope != null, "Scope is null.")

  def core() : Schedulable = container() match {
    case x:ServerLike => x.core()
    case _ => null
  }

  def container() = _theScope.container()
  def isActive() = _active
  def job() = _theScope
  def getPID() = _pid

  protected def onEnd() {
    _delegate.onStop(this)
  }

  protected[wflow] def onError(e:Throwable, cur:FlowPoint) : Activity = {
    tlog().error("", e)
    _delegate.onError(e,cur) match {
      case x:Activity => x
      case _ => new Nihil()
    }
  }

  protected def onStart():Activity = {
    _delegate.getStartActivity(this) match {
      case x:Activity => x
      case _ => new Nihil()
    }
  }

  def start() {
    tlog().debug("{}: {} => pid : {} => starting" , "Pipeline", _delegateClass , asJObj(_pid))

    try {
      val f1= onStart().reify( new NihilPoint(this))
      _active=true
      core().run(f1)
    } catch {
      case e:Throwable =>
        tlog().error("", e)
        stop()
    }
  }

  def stop() {
    try {
      onEnd()
    } catch {
      case e:Throwable =>
      tlog().error("",e)
    }

    tlog().debug("{}: {} => pid : {} => end" , "Pipeline", _delegateClass , asJObj(_pid))
  }

  override def toString() = {
    _delegateClass + "(" + _pid + ")"
  }

//  override def finalize() {
//    super.finalize()
//    println("=========================> Pipeline: " + getClass.getName + " finz'ed")
//  }
  
}
